File: aemet/SergioDeIscar-aemet-java/SergioDeIscar-aemet-java/src/main/java/org/example/repositories/dupla/DuplaRepositoryMap.java
package org.example.repositories.dupla;

import org.example.models.Dupla;
import org.example.services.storage.dupla.DuplaStorageService;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.*;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.mapping;

public class DuplaRepositoryMap implements DuplaRepository{
    private final DuplaStorageService storageService;
    public DuplaRepositoryMap(DuplaStorageService storageService) {
        this.storageService = storageService;
    }

    //private final Map<LocalDate, Map<String, Dupla>> duplas = new LinkedHashMap<>();
    private final Map<String, Dupla> duplas = new LinkedHashMap<>();

    @Override
    public List<Dupla> getAll() {
        upgrade();
        return duplas.values().stream().toList();
    }

    @Override
    public Dupla getById(String s) {
        upgrade();
        return duplas.get(s);
    }

    @Override
    public Dupla save(Dupla element, boolean storage) {
        duplas.put(element.getPoblacion() + " " + element.getDay(), element);
        if (storage) downgrade();
        return element;
    }

    @Override
    public void saveAll(List<Dupla> elements, boolean storage) {
        elements.forEach(e -> save(e, storage));
    }

    @Override
    public Dupla deleteById(String s) {
        var element = duplas.get(s);
        if (element == null) return null;
        duplas.remove(s);
        downgrade();
        return element;
    }

    @Override
    public Dupla update(Dupla element) {
        return updateById(element.getPoblacion() + " " + element.getDay(), element);
    }

    @Override
    public Dupla updateById(String s, Dupla element) {
        if (duplas.get(s) == null) return null;
        duplas.put(s, element);
        downgrade();
        return element;
    }

    @Override
    public List<Dupla> upgrade() {
        duplas.clear();
        var load = storageService.loadAll();
        saveAll(load, false);
        return load;
    }

    @Override
    public List<Dupla> downgrade() {
        return storageService.saveAll(duplas.values().stream().toList());
    }

    @Override
    public Map<LocalDate, Map<String, Dupla>> maxTemPorLugar() {
        upgrade();
        Map<LocalDate, Map<String, Optional<Dupla>>> optional = duplas.values().stream()
                .collect(groupingBy(
                        Dupla::getDay,
                        groupingBy(
                                Dupla::getPoblacion,
                                Collectors.maxBy(Comparator.comparing(Dupla::getTemMax))
                        ))
                );
        // Cambiar Optional<Dupla> por Dupla
        return removeOptional(optional);
        // Java es peor que Kotlin en este aspecto
    }

    @Override
    public Map<LocalDate, Map<String, Dupla>> minTemPorLugar() {
        upgrade();
        Map<LocalDate, Map<String, Optional<Dupla>>> optional = duplas.values().stream()
                .collect(groupingBy(
                        Dupla::getDay,
                        groupingBy(
                                Dupla::getPoblacion,
                                Collectors.minBy(Comparator.comparing(Dupla::getTemMin))
                        ))
                );
        // Cambiar Optional<Dupla> por Dupla
        return removeOptional(optional);
    }

    private Map<LocalDate, Map<String, Dupla>> removeOptional(Map<LocalDate, Map<String, Optional<Dupla>>> optional) {
        return optional.entrySet().stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey, e -> e.getValue().entrySet().stream()
                                .collect(Collectors.toMap(
                                        Map.Entry::getKey,
                                        e2 -> e2.getValue().get()
                                ))
                ));
    }

    @Override
    public Map<String, Dupla> maxTemPorProvincia() {
        upgrade();
        return duplas.values().stream()
                .collect(Collectors.toMap(
                        Dupla::getProvincia,
                        e -> e,
                        (e1, e2) -> e1.getTemMax() > e2.getTemMax() ? e1 : e2
                ));
    }

    @Override
    public Map<String, Dupla> minTemPorProvincia() {
        upgrade();
        return duplas.values().stream()
                .collect(Collectors.toMap(
                        Dupla::getProvincia,
                        e -> e,
                        (e1, e2) -> e1.getTemMin() < e2.getTemMin() ? e1 : e2
                ));
    }

    @Override
    public Map<String, Double> mediaPorProvincia() {
        upgrade();
        return duplas.values().stream()
                .collect(Collectors.groupingBy(
                        Dupla::getProvincia,
                        Collectors.averagingDouble(e -> e.getTemMax() - e.getTemMin())
                ));
    }

    @Override
    public Map<LocalDate, Map<String, Double>> mediaPrecipitacionPorDiaYProvincia() {
        upgrade();
        return duplas.values().stream()
                .collect(groupingBy(
                        Dupla::getDay,
                        groupingBy(
                                Dupla::getProvincia,
                                Collectors.averagingDouble(Dupla::getPrecipitacion)
                        )
                ));
    }

    @Override
    public Double mediaTemMadrid() {
        upgrade();
        return duplas.values().stream()
                .filter(e -> e.getProvincia().equals("Madrid"))
                .mapToDouble(e -> e.getTemMax() - e.getTemMin())
                .average()
                .orElse(0.0);
    }

    @Override
    public Double mediaTemMax() {
        upgrade();
        return duplas.values().stream()
                .mapToDouble(Dupla::getTemMax)
                .average()
                .orElse(0.0);
    }

    @Override
    public Double mediaTemMin() {
        upgrade();
        return duplas.values().stream()
                .mapToDouble(Dupla::getTemMin)
                .average()
                .orElse(0.0);
    }

    @Override
    public Map<LocalDate, List<String>> maxTemAntes() {
        upgrade();
        return duplas.values().stream()
                .filter(e -> e.getTimeMax().isBefore(LocalTime.of(15, 0)))
                .collect(groupingBy(
                        Dupla::getDay,
                        mapping(Dupla::getPoblacion, Collectors.toList())
                ));
    }

    @Override
    public Map<LocalDate, List<String>> minTemDespues() {
        upgrade();
        return duplas.values().stream()
                .filter(e -> e.getTimeMin().isAfter(LocalTime.of(17, 30)))
                .collect(groupingBy(
                        Dupla::getDay,
                        mapping(Dupla::getPoblacion, Collectors.toList())
                ));
    }

    @Override
    public Dupla maxTemMadrid(LocalDate day) {
        upgrade();
        return duplas.values().stream()
                .filter(e -> e.getPoblacion().equals("Madrid") && e.getDay().equals(day))
                .max(Comparator.comparing(Dupla::getTemMax))
                .orElse(null);
    }

    @Override
    public Dupla minTemMadrid(LocalDate day) {
        upgrade();
        return duplas.values().stream()
                .filter(e -> e.getPoblacion().equals("Madrid") && e.getDay().equals(day))
                .min(Comparator.comparing(Dupla::getTemMin))
                .orElse(null);
    }

    @Override
    public Double mediaPrecipitacionMadrid(LocalDate day) {
        upgrade();
        return duplas.values().stream()
                .filter(e -> e.getPoblacion().equals("Madrid") && e.getDay().equals(day))
                .mapToDouble(Dupla::getPrecipitacion)
                .average()
                .orElse(0.0);
    }
}

File: aemet/SergioDeIscar-aemet-kotlin/SergioDeIscar-aemet-kotlin/src/main/kotlin/repositories/dupla/DuplaRepositoryMap.kt
package repositories.dupla

import models.Dupla
import models.Informe
import mu.KotlinLogging
import services.storage.dupla.DuplaStorageService
import java.time.LocalDate
import java.time.LocalTime

private val logger = KotlinLogging.logger {  }

class DuplaRepositoryMap(
    private val storageService: DuplaStorageService
): DuplaRepository {
    private val duplas = mutableMapOf<String,Dupla>()

    override fun maxTemPorLugar(): Map<LocalDate, Map<String, Double>> {
        logger.debug { "DuplaRepositoryMap ->\tmaxTemPorLugar" }
        upgrade()
        return duplas.values
            .groupBy { it.day }
            .mapValues {
                it.value.groupBy { it.poblacion }
                    .mapValues {
                        it.value.maxBy { it.temMax }.temMax
                    }
            }
    }

    override fun minTemPorLugar(): Map<LocalDate, Map<String, Double>> {
        logger.debug { "DuplaRepositoryMap ->\tminTemPorLugar" }
        upgrade()
        return duplas.values
            .groupBy { it.day }
            .mapValues {
                it.value.groupBy { it.poblacion }
                    .mapValues {
                        it.value.minBy { it.temMin }.temMin
                    }
            }
    }

    override fun maxTemPorProvincia(): Map<String, Double> {
        logger.debug { "DuplaRepositoryMap ->\tmaxTemPorProvincia" }
        upgrade()
        return duplas.values
            .groupBy { it.provincia }
            .mapValues {
                it.value.maxBy { it.temMax }.temMax
            }
    }

    override fun minTemPorProvincia(): Map<String, Double> {
        logger.debug { "DuplaRepositoryMap ->\tminTemPorProvincia" }
        upgrade()
        return duplas.values
            .groupBy { it.provincia }
            .mapValues {
                it.value.minBy { it.temMin }.temMin
            }
    }

    override fun mediaPorProvincia(): Map<String, Double> {
        logger.debug { "DuplaRepositoryMap ->\tmediaPorProvincia" }
        upgrade()
        return duplas.values
            .groupBy { it.provincia }
            .mapValues {
                it.value.map {
                    it.temMax - it.temMin
                }.average()
            }
    }

    override fun mediaPrecipitacionPorDiaYProvincia(): Map<LocalDate, Map<String, Double>> {
        logger.debug { "DuplaRepositoryMap ->\tmediaPrecipitacionPorDiaYProvincia" }
        upgrade()
        return duplas.values
            .groupBy { it.day }
            .mapValues {
                it.value.groupBy { it.provincia }
                    .mapValues {
                        it.value.map { it.precipitacion }.average()
                    }
            }
    }

    override fun mediaTemMadrid(): Double {
        logger.debug { "DuplaRepositoryMap ->\tmediaTemMadrid" }
        upgrade()
        return duplas.values
            .filter { it.provincia == "Madrid" }
            .map { it.temMax - it.temMin }
            .average()
    }

    override fun mediaTemMax(): Double {
        logger.debug { "DuplaRepositoryMap ->\tmediaTemMax" }
        upgrade()
        return duplas.values
            .map { it.temMax }
            .average()
    }

    override fun mediaTemMin(): Double {
        logger.debug { "DuplaRepositoryMap ->\tmediaTemMin" }
        upgrade()
        return duplas.values
            .map { it.temMin }
            .average()
    }

    override fun maxTemAntes(): Map<LocalDate, List<String>> {
        logger.debug { "DuplaRepositoryMap ->\tmaxTemAntes" }
        upgrade()
        return duplas.values
            .groupBy { it.day }
            .mapValues {
                it.value.filter { it.timeMax.isBefore(LocalTime.of(15, 0)) }
                    .map { it.poblacion }
            }
    }

    override fun minTemDespues(): Map<LocalDate, List<String>> {
        logger.debug { "DuplaRepositoryMap ->\tminTemDespues" }
        upgrade()
        return duplas.values
            .groupBy { it.day }
            .mapValues {
                it.value.filter { it.timeMax.isAfter(LocalTime.of(17, 30)) }
                    .map { it.poblacion }
            }
    }

    override fun maxTemMadrid(day: LocalDate): Dupla {
        logger.debug { "DuplaRepositoryMap ->\tmaxTemMadrid" }
        upgrade()
        return duplas.values
            .filter { it.provincia == "Madrid" && it.day == day }
            .maxBy { it.temMax }
    }

    override fun minTemMadrid(day: LocalDate): Dupla {
        logger.debug { "DuplaRepositoryMap ->\tminTemMadrid" }
        upgrade()
        return duplas.values
            .filter { it.provincia == "Madrid" && it.day == day }
            .minBy { it.temMin }
    }

    override fun mediaPrecipitacionMadrid(day: LocalDate): Double {
        logger.debug { "DuplaRepositoryMap ->\tmediaPrecipitacionMadrid" }
        upgrade()
        return duplas.values
            .filter { it.provincia == "Madrid" && it.day == day }
            .map { it.precipitacion }
            .average()
    }

    override fun getAll(): List<Dupla> {
        logger.debug { "DuplaRepositoryMap ->\tgetAll" }
        upgrade()
        return duplas.values.toList()
    }

    override fun getById(id: String): Dupla? {
        logger.debug { "DuplaRepositoryMap ->\tgetById" }
        upgrade()
        return duplas.values.find { it.poblacion == id }
    }

    override fun save(element: Dupla, storage: Boolean): Dupla {
        logger.debug { "DuplaRepositoryMap ->\tsave" }
        duplas["${element.poblacion} ${element.day}"] = element
        if (storage) downgrade()
        return element
    }

    override fun saveAll(elements: List<Dupla>, storage: Boolean) {
        logger.debug { "DuplaRepositoryMap ->\tsaveAll" }
        elements.forEach { save(it, storage) }
    }

    override fun deleteById(id: String): Dupla? {
        logger.debug { "DuplaRepositoryMap ->\tdeleteById" }
        val dupla = getById(id) ?: return null
        duplas.remove("${dupla.poblacion} ${dupla.day}")
        downgrade()
        return dupla
    }

    override fun update(element: Dupla): Dupla? {
        logger.debug { "DuplaRepositoryMap ->\tupdate" }
        return updateById(element.poblacion, element)
    }

    override fun updateById(id: String, element: Dupla): Dupla? {
        logger.debug { "DuplaRepositoryMap ->\tupdateById" }
        deleteById(id) ?: return null
        return save(element)
    }

    override fun upgrade(): List<Dupla> {
        logger.debug { "DuplaRepositoryMap ->\tupgrade" }
        duplas.clear()
        val load = storageService.loadAll()
        saveAll(load, false)
        return load
    }

    override fun downgrade(): List<Dupla> {
        logger.debug { "DuplaRepositoryMap ->\tdowngrade" }
        return storageService.saveAll(duplas.values.toList())
    }
}